docs/polkadot/implementors-guide/guide.md

```diff
@@ -619,7 +619,7 @@ Validator group assignments do not need to change very quickly. The security ben
 
 Validator groups rotate across execution cores in a round-robin fashion, with rotation occurring at fixed intervals. The i'th group will be assigned to the `(i+k)%n`'th core at any point in time, where `k` is the number of rotations that have occurred in the session, and `n` is the number of cores. This makes upcoming rotations within the same session predictable.
 
-When a rotation occurs, validator groups are still responsible for distributing availability pieces for any previous cores that are still occupied and pending availability. In practice, rotation and availability-timeout frequencies should be set so this will only be the core they have just been rotated from. It is possible that a validator group is rotated onto a core which is currently occupied. In this case, the validator group will have nothing to do until the previously-assigned group finishes their availability work and frees the core or the availability process times out. Depending on if the core is for a parachain or parathread, a different timeout `t` from the `HostConfiguration` will apply. Availability timeouts should only be triggered in the first `t-1` blocks after the beginning of a session.
+When a rotation occurs, validator groups are still responsible for distributing availability pieces for any previous cores that are still occupied and pending availability. In practice, rotation and availability-timeout frequencies should be set so this will only be the core they have just been rotated from. It is possible that a validator group is rotated onto a core which is currently occupied. In this case, the validator group will have nothing to do until the previously-assigned group finishes their availability work and frees the core or the availability process times out. Depending on if the core is for a parachain or parathread, a different timeout `t` from the `HostConfiguration` will apply. Availability timeouts should only be triggered in the first `t-1` blocks after the beginning of a rotation.
```

Comment:
Is there a missing "in practice" the last sentence?  We could've a validator group with enough crappy validators who fail (a) distributing pieces, or (b) providing the block to stronger validator group members, but a subsequent luckier collator have more luck reaching the stronger validator group members.

Reply:
This was Al's suggestion. If you have enough honest validators in your group to back a candidate, you shouldn't have any trouble distributing pieces, right?

Reply:
I'll ask him.  I think dishonest or crappy validators could halt the parachain for one rotation this way.

Reply:
If the group has enough crappy validators, then it can always halt block production for a rotation. The goal here was to try to ensure that they couldn't tie it up for any more than a rotation by getting the new validators to validate as fast as possible.

We need at least of majority of a group to back a candidate; I don't know if we ever decided on the exact number here. But if any one of these is competetent, they should be able to do (b), whereas (a) might need several competent validators. 

So we should be able to assume that the group collectively is crappy and try to get the next guys on the job as fast as possible.

Reply:
We should focus our efforts on dealing with this kind of situation if it turns out to happen frequently by fixing networking on the validator and collator side. Timing out availability might just give us the same problem on the next block, since it is the same group of validators and the backing validators will overlap a lot.

Reply:
Is there any code simplification or other benefit in constraining when availability timeouts occur though?  

At first I'd imagined some high proportion like 2/3 solves my concern, but If all but one validator turn out crappy then allowing later timeouts improves performance. 

We've mostly discussed 1/2 but actually 2/3 of the group ensures that 2/3 of validators can contact the group, assuming no NATs, etc.  I think high like 2/3 ensure availability runs faster, while lower makes availability start slower, but actually gives availability more running time.

Reply:
It's slightly cheaper to avoid doing a GC sweep every block. We're currently doing this in an O(N) way, but we could also trade off more storage and build an index for knowing better which cores need to be collected.

I lean towards the fewer timeouts approach for that reason and the fact that there are other ways to halt parachain inclusion for a rotation. Rotations are relatively fast, so only timing out recently after rotations wouldn't be a huge issue.

I've been assuming a simple majority, not a supermajority for candidate groups, because liveness is probably a bigger issue than safety in the early versions of Polkadot. We can tweak this later if we want.